### PR TITLE
feature/add pundit, create seeds and seat validation

### PR DESCRIPTION
**Pull request by atytko**
- Created seeds
- Added pundit and basic authorization for existing controllers
- Created seat validations and specs
- Created reservations controller
- Fixed failing request specs (I will create more specs checking authorisation in a separate PR, since I have already flooded this PR 👀 )

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -10,3 +10,112 @@
...
+  password: 'password'
+)
+
+User.create!(
+  email: 'employee@test.com',
+  first_name: 'Anna',
+  last_name: 'Smith',
+  password: 'password'
+)
+User.last.update(user_role: UserRole.find_by(name: 'employee'))
```
Comment (w0st) [inline db/seeds.rb:28]:
```suggestion
User.create!(
  email: 'employee@test.com',
  first_name: 'Anna',
  last_name: 'Smith',
  password: 'password',
  user_role: UserRole.find_by(name: 'employee')
)
```
?

File: db/seeds.rb [hunk 2]
```diff
@@ -10,3 +10,112 @@
...
+              ])
+Rails.logger.info 'Movies created.'
+
+Rails.logger.info 'Creating Ticket Desks...'
+ticket_desk = TicketDesk.create!
+Rails.logger.info 'Ticket Desks created.'
+
+Rails.logger.info 'Creating Screenings...'
+Screening.create!([
+                    { cinema_hall: CinemaHall.all.sample, screening_time: DateTime.now + 1.hour,
```
Comment (w0st) [inline db/seeds.rb:92]: Personally, I always try to use `DateTime.current` (or even `Time.current`) over `DateTime.now`. You can find here a nice explanation  of why:
https://thoughtbot.com/blog/its-about-time-zones

**Feel free to ignore the comment at this moment, but definitely find a time to read the article.** You can avoid potential problems in future :wink:

File: db/seeds.rb [hunk 3]
```diff
@@ -10,3 +10,112 @@
...
+                      { status: 'pending', screening: Screening.all.sample, ticket_desk: ticket_desk, user: client  },
+                      { status: 'pending', screening: Screening.all.sample, ticket_desk: ticket_desk, user: client  }
+                    ])
+Rails.logger.info 'Reservations created.'
+
+Rails.logger.info 'Creating Seats Reservations...'
+letters = ('A'..'J').to_a
+Reservation.all.each do |reservation|
+  rand(1..3).times do |_number|
+    SeatsReservation.create!([row: letters[rand(1..reservation.screening.cinema_hall.row_number) - 1],
```
Comment (w0st) [inline db/seeds.rb:115]: If you're looking for an easy option to get random numbers from a predefined set without repeating you can try [this simple solution](https://stackoverflow.com/questions/9457841/generate-non-repeating-random-numbers-in-ruby#answer-9458016).

File: spec/validators/seat_validator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    context 'when seat reservation already exists for this screening' do
+      let!(:seats_reservation) do
+        create(:seats_reservation, reservation: reservation,
+                                   cinema_hall: cinema_hall, row: row, seat_number: seat_number)
+      end
+
+      it { is_expected.to eq(true) }
+    end
+
+    it {}
```
Comment (w0st) [inline spec/validators/seat_validator_spec.rb:36]: Would you like to add another test to this file? Currently, you have an empty "it".

File: app/models/seats_reservation.rb [hunk 5]
```diff
@@ -3,4 +3,22 @@
...
+        seat_number: seat_number,
+        reservations: { screening: reservation.screening }
+      )
+      .any?
+  end
+
+  private
+
+  def validates_seat_reservation
+    errors.add(:seat_reservation, 'this seat is already taken') if seat_is_taken?
```
Comment (w0st) [inline app/models/seats_reservation.rb:22]: We discussed that this validation doesn't make sense at all at the model level. Let me explain why.

0. (initial state). The system doesn't have any `SeatReservation` records.
1. We're creating a new object of `SeatReservation`:  `seat_reservation_1 = SeatReservation.new(row: 'A', seat_number: 6, cinema_hall: cinema_hall, reservation: reservation)`.
2. We're checking this validation: `seat_reservation_1.valid? # => true`
3. We're saving the record: `seat_reservation_1.save`
4. We're checking the validation again: `seat_reservation_1.valid? # => false`
I don't think that you would like to achieve a behaviour like that. You basically will end up always with an invalidated record in the database after saving it to the database. I think that you would like to perform this "checking" before saving in your business process instead of treating it as an invariant.

https://lostechies.com/jimmybogard/2016/04/29/validation-inside-or-outside-entities/ may this will be helpful for you.

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -2,8 +2,16 @@
...
   include DeviseTokenAuth::Concerns::SetUserByToken
+  include Pundit
   before_action :configure_permitted_parameters, if: :devise_controller?
   before_action :authenticate_user!
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+
+  private
+
+  def user_not_authorized
+    render json: { error: 'You are not authorized to perform this action.' }
```
Comment (w0st) [inline app/controllers/application_controller.rb:13]: The HTTP has already defined [codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) for some typical situations. In current shape, the server returns `200` (its the default value) which means "The request has succeeded." which is a lie. The `render` has a possibility to set the HTTP code for a response by `status` parameter. Take a look at the description in [the Rails guide](https://guides.rubyonrails.org/v2.3/layouts_and_rendering.html#the-status-option), please.

For me, the https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403 seems to be the most appropriate in the case.

File: app/controllers/user_reservations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class UserReservationsController < ApplicationController
+  include JSONAPI::Fetching
+
+  def index
+    reservations = current_user.reservations
+
```
Comment (w0st) [inline app/controllers/user_reservations_controller.rb:8]: What about setting a pundit policy here?  :smile:

File: app/policies/application_policy.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+class ApplicationPolicy
+  attr_reader :user, :record
+
+  def initialize(user, record)
+    @user = user
+    @record = record
+  end
+
+  def index?
```
Comment (w0st) [inline app/policies/application_policy.rb:11]: I think you can simplify these policy classes a lot - in most cases, the rules are the same for all methods. I don't want to block the PR. Try to express the rules more briefly, please. In that way, you will help someone (and especially yourself) in the future with understanding this part of code.

**Review by w0st — CHANGES_REQUESTED** [review]
The current validation rule in the model will lead you to many strange "errors" - you need to correct them.

Consider the rest of my comments, but feel free to ignore them at this stage :wink:

File: db/seeds.rb (hunk 1 above)
Comment (atytko) [reply db/seeds.rb:28]: So this was my original version, but then I started to wonder if this wouldn't be a security issue? My idea behind permissions in this app is that user sign ups as a regular user and then is given permission of employee or manager if applicable (so that there is no risk of a situation in which a potential hacker assign himself in the sign up form a role of employee or manager. In the suggested change, wouldn't this be a problem or this is just seeds and I shouldn't worry about it at all? :D

File: db/seeds.rb (hunk 2 above)
Comment (atytko) [reply db/seeds.rb:92]: Thanks for the article, will have a look 👍  Also: I updated it to DateTime.current. :)

File: spec/validators/seat_validator_spec.rb (hunk 4 above)
Comment (atytko) [reply spec/validators/seat_validator_spec.rb]: I removed it, thanks for spotting it 👍

File: app/controllers/user_reservations_controller.rb (hunk 7 above)
Comment (atytko) [reply app/controllers/user_reservations_controller.rb:8]: Will the `current_user` not do the job? 🤔 I have added this policy in the index method inside employee reservations controller though.

File: app/models/seats_reservation.rb (hunk 5 above)
Comment (atytko) [reply app/models/seats_reservation.rb:24]: Sure, I remember, but I just had some issue with using method` seat_is_taken?` inside my reservations controller, so I decided to create a draft PR for now and reach out for some help on Monday :)

File: app/controllers/application_controller.rb (hunk 6 above)
Comment (atytko) [reply app/controllers/application_controller.rb]: Makes sense, thanks 👍  I set the status as forbidden.

File: db/seeds.rb (hunk 1 above)
Comment (w0st) [reply db/seeds.rb:28]: It sounds good. Both approaches work for me.

File: app/controllers/user_reservations_controller.rb (hunk 7 above)
Comment (w0st) [reply app/controllers/user_reservations_controller.rb:8]: Indeed it should be fine :ok_hand:

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -2,8 +2,16 @@
...
   include DeviseTokenAuth::Concerns::SetUserByToken
+  include Pundit
   before_action :configure_permitted_parameters, if: :devise_controller?
   before_action :authenticate_user!
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+
+  private
+
+  def user_not_authorized
+    render status: :forbidden
```
Comment (w0st) [inline app/controllers/application_controller.rb:13]: I had in mind this:
```suggestion
    render json: { error: 'You are not authorized to perform this action.' }, status: :forbidden
```

File: app/policies/application_policy.rb (hunk 8 above)
Comment (atytko) [reply app/policies/application_policy.rb:11]: I have refactored policies, thanks.

File: app/models/seats_reservation.rb (hunk 5 above)
Comment (atytko) [reply app/models/seats_reservation.rb]: Refactored, thanks for your help 👍

File: app/policies/application_policy.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    false
+  end
+
+  def create?
+    false
+  end
+
+  def new?
+    create?
+  end
```
Comment (rafalrothenberger) [inline app/policies/application_policy.rb:25]: As we are doing API only application we will not have in our controllers actions for `new` and `edit` - only for `update` and `create`, so you can drop both from your policies

File: app/controllers/application_controller.rb (hunk 9 above)
Comment (rafalrothenberger) [reply app/controllers/application_controller.rb:13]: It would be best to have both error and 403 as HTTP Code as Wojtek is proposing

File: app/controllers/user_reservations_controller.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+class UserReservationsController < ApplicationController
+  include JSONAPI::Fetching
+
+  def index
+    reservations = current_user.reservations
```
Comment (rafalrothenberger) [inline app/controllers/user_reservations_controller.rb:7]: It would be great to add policy from pundit here, it would allow you to have a single controller for both employees and users - normal user would get through that policy only their own reservations and employee would get all of them:
https://github.com/varvet/pundit#scopes

File: app/controllers/user_reservations_controller.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+    render jsonapi: reservations, except: blacklisted_attributes
+  end
+
+  def show
+    reservation = current_user.reservations.find(params[:id])
```
Comment (rafalrothenberger) [inline app/controllers/user_reservations_controller.rb:13]: You can handle that through pundit as well - again hawing single controller instead of two. In the Policy, you would return `true` always for employees, for users, you would return true only if `user.id == reservations.user_id` - that way users would have access only to their own reservations and employees to all of them.

**Review by rafalrothenberger — COMMENTED** [review]
So in theory nothing blocking, but I would argue that following that advice about merging `ReservationController` and `EmployeeReservationController` and handling accesses through pundit instead would be a good idea.

File: app/controllers/user_reservations_controller.rb (hunk 11 above)
Comment (atytko) [reply app/controllers/user_reservations_controller.rb]: I merged controllers and refactored policies, thanks.

File: app/controllers/reservations_controller.rb [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  include JSONAPI::Fetching
+
+  def index
+    reservations = policy_scope(Reservation).where(screening: params[:screening_id])
+
+    render jsonapi: reservations, except: blacklisted_attributes
+  end
+
+  def show
+    reservation = policy_scope(Reservation).find(params[:id])
```
Comment (rafalrothenberger) [inline app/controllers/reservations_controller.rb:13]: Ok, for show it would be better to do
```ruby
reservation = Resevation.find(params[:id])
authorize reservation
```
That way, when accessing a reservation that I do not have access to, it will return an unauthorized error from pundit and that will be translated to correct unauthorized message in your API. Your current code would return a 404 error instead.

File: app/controllers/reservations_controller.rb [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+    render jsonapi: reservation, except: blacklisted_attributes
+  end
+
+  def create
+    if already_booked?
+      render json: { error: 'One or more seats is taken' }, status: :unprocessable_entity
+    else
+      reservation = current_user.reservations.new(reservation_params)
+      authorize reservation
```
Comment (rafalrothenberger) [inline app/controllers/reservations_controller.rb:23]: So in the case of creating it is, in most cases, better to have `authorize Reservation` at the beginning of the function instead of authorizing whenever someone is able to create this exact resource.

File: app/policies/reservation_policy.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+      user.id == record.user_id
+    end
+  end
+
+  def create?
+    true
+  end
+
+  def update?
+    true
```
Comment (rafalrothenberger) [inline app/policies/reservation_policy.rb:27]: It would be best to have the same logic here as in show? - you can extract it to some function like `admin_or_owner?` and call it in both.

File: db/seeds.rb [hunk 16]
```diff
@@ -10,3 +10,112 @@
...
+                              seat_number: rand(1..reservation.screening.cinema_hall.row_total_seats),
+                              cinema_hall: reservation.screening.cinema_hall,
+                              reservation: reservation])
```
Comment (rafalrothenberger) [inline db/seeds.rb:118]: Ok so, the chance for that is low, but if we would run seeds in CI then it will start happening because of the number of repetitions of that process.

You are using rand which returns a random value - as such there is really nothing preventing it to return the same row and seat_number pair twice in that 3 repetitions.

Wojciech pointed you to shuffle method, which could be used like that:

```suggestion
rows= ('A'..'J').to_a.shuffle
seats=(1..reservation.screening.cinema_hall.row_total_seats).to_a.shuffle
Reservation.all.each do |reservation|
  rand(1..3).times do |number|
    SeatsReservation.create!([row: rows[number],
                              seat_number: seats[number],
                              cinema_hall: reservation.screening.cinema_hall,
                              reservation: reservation])
```

As there are no repetitions in `seats` and `rows` array it will never return twice the same pair in that 3 repetitions.

File: app/policies/reservation_policy.rb (hunk 15 above)
Comment (atytko) [reply app/policies/reservation_policy.rb:27]: Refactored, thanks :)

File: db/seeds.rb (hunk 16 above)
Comment (atytko) [reply db/seeds.rb:118]: Implemented as discuessed:
`Reservation.all.each do |reservation|
  rows = ('A'..'J').to_a.shuffle
  seats = (1..reservation.screening.cinema_hall.row_total_seats).to_a.shuffle
  rand(1..3).times do |number|
    SeatsReservation.create!([row: rows[number],
                              seat_number: seats[number],
                              cinema_hall: reservation.screening.cinema_hall,
                              reservation: reservation])
  end
end`

File: spec/requests/cinema_halls_controller_spec.rb [hunk 17]
```diff
@@ -13,6 +13,9 @@
     let(:cinema_hall) { create(:cinema_hall) }
 
     context 'when the user is logged in' do
+      let(:role) { create(:user_role, name: 'employee') }
+      let(:role) { create(:user_role, name: 'manager') }
```
Comment (rafalrothenberger) [inline spec/requests/cinema_halls_controller_spec.rb:17]: You often has it twice in your specs, probably should just left the one you want :)

Comment (atytko) [reply spec/requests/cinema_halls_controller_spec.rb]: Refactored, thanks :)

**Review by rafalrothenberger — APPROVED** [review]
🎉 Look good, great job!
